File: backend/src/main/java/com/autocare/mechanic/controller/MechanicController.java
package com.autocare.mechanic.controller;

import com.autocare.mechanic.dto.MechanicRequestDTO;
import com.autocare.mechanic.dto.MechanicResponseDTO;
import com.autocare.mechanic.service.MechanicService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * Gestão da equipe (cadastro de mecânicos) é responsabilidade de ADMIN/MANAGER.
 * Leitura é liberada também para RECEPTIONIST, que precisa consultar mecânicos
 * disponíveis para atribuir a uma ordem de serviço.
 */
@RestController
@RequestMapping("/mechanics")
@RequiredArgsConstructor
@PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'RECEPTIONIST')")
public class MechanicController {

    private final MechanicService mechanicService;

    @PostMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<MechanicResponseDTO> create(@Valid @RequestBody MechanicRequestDTO request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(mechanicService.create(request));
    }

    @GetMapping
    public ResponseEntity<List<MechanicResponseDTO>> findAll() {
        return ResponseEntity.ok(mechanicService.findAll());
    }

    @GetMapping("/available")
    public ResponseEntity<List<MechanicResponseDTO>> findAvailable() {
        return ResponseEntity.ok(mechanicService.findAvailable());
    }

    @GetMapping("/{id}")
    public ResponseEntity<MechanicResponseDTO> findById(@PathVariable String id) {
        return ResponseEntity.ok(mechanicService.findByIdResponse(id));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<MechanicResponseDTO> update(
            @PathVariable String id,
            @Valid @RequestBody MechanicRequestDTO request) {
        return ResponseEntity.ok(mechanicService.update(id, request));
    }

    @PatchMapping("/{id}/availability")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<Void> setAvailability(
            @PathVariable String id,
            @RequestParam boolean available) {
        mechanicService.setAvailability(id, available);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> delete(@PathVariable String id) {
        mechanicService.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend/src/mechanic/mechanic.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseBoolPipe,
  Patch,
  Post,
  Put,
  Query,
  UseGuards,
} from "@nestjs/common";
import { MechanicService } from "./mechanic.service";
import { MechanicRequestDto } from "./dto/mechanic.dto";
import { Roles } from "../common/decorators/roles.decorator";
import { RolesGuard } from "../common/guards/roles.guard";
import { Role } from "@prisma/client";

// Gestão da equipe (cadastro de mecânicos) é responsabilidade de ADMIN/MANAGER.
// Leitura é liberada também para RECEPTIONIST, que precisa consultar mecânicos
// disponíveis para atribuir a uma ordem de serviço.
@Controller("mechanics")
@UseGuards(RolesGuard)
@Roles(Role.ADMIN, Role.MANAGER, Role.RECEPTIONIST)
export class MechanicController {
  constructor(private readonly mechanicService: MechanicService) {}

  @Post()
  @Roles(Role.ADMIN, Role.MANAGER)
  @HttpCode(HttpStatus.CREATED)
  create(@Body() dto: MechanicRequestDto) {
    return this.mechanicService.create(dto);
  }

  @Get()
  findAll() {
    return this.mechanicService.findAll();
  }

  @Get("available")
  findAvailable() {
    return this.mechanicService.findAvailable();
  }

  @Get(":id")
  findById(@Param("id") id: string) {
    return this.mechanicService.findByIdResponse(id);
  }

  @Put(":id")
  @Roles(Role.ADMIN, Role.MANAGER)
  update(@Param("id") id: string, @Body() dto: MechanicRequestDto) {
    return this.mechanicService.update(id, dto);
  }

  @Patch(":id/availability")
  @Roles(Role.ADMIN, Role.MANAGER)
  @HttpCode(HttpStatus.NO_CONTENT)
  async setAvailability(@Param("id") id: string, @Query("available", ParseBoolPipe) available: boolean) {
    await this.mechanicService.setAvailability(id, available);
  }

  @Delete(":id")
  @Roles(Role.ADMIN, Role.MANAGER)
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param("id") id: string) {
    await this.mechanicService.delete(id);
  }
}
